### PR TITLE
Big refactor of exporter.

### PR DESCRIPTION
**Pull request by Yannig**
Refactor the way we retrieve informations from oracle by introducing new
functions ScrapeGenericValues and GeneratePrometheusMetrics to simplify
Prometheus metrics generation.

Porting several existing function to use ScrapeGenericValues and removing
specific methods to scrape values.

Introducing new Metric struct describing metrics to scrape.
Add a new structure with all metrics defined.

New option to add custom metrics from TOML file. Add examples and update
README file. You can also use env var.

Stop using a new connection everytime we scrape values. Go handle case where
connections is lost.

New metrics to retrieve asm diskgroup informations and resource limit.

### CODE REVIEW

**Comment by greenpau** [comment]
@Yannig , what is the status of this refactoring?

**Comment by greenpau** [comment]
At the moment, CircleCI fails because it cannot find `github.com/BurntSushi/toml`.

```
main.go:13:2: cannot find package "github.com/BurntSushi/toml" in any of:
	/home/travis/gopath/src/github.com/iamseth/oracledb_exporter/vendor/github.com/BurntSushi/toml (vendor tree)
	/home/travis/.gimme/versions/go1.9.linux.amd64/src/github.com/BurntSushi/toml (from $GOROOT)
	/home/travis/gopath/src/github.com/BurntSushi/toml (from $GOPATH)
```

I think `go get` around https://github.com/iamseth/oracledb_exporter/blob/master/Makefile#L22 might help 😄

**Comment by Yannig** [comment]
@greenpau thanks for your feedback. Actually, this refactoring is working for me so I'm using it on daily basis.

There's also another change I have made on my side: timeout on oracle query. Let me known if you are interested.

I'll take a look to add go get to run Travis.

File: README.md [hunk 1]
```diff
@@ -31,6 +31,16 @@ The following metrics are exposed currently.
 - oracledb_tablespace_bytes
 - oracledb_tablespace_max_bytes
 - oracledb_tablespace_bytes_free
+- oracledb_resource_current_utilization
+- oracledb_resource_limit_value
+- oracledb_wait_time_application
+- oracledb_wait_time_commit
```
Comment (greenpau) [inline README.md]: @Yannig , this is duplicate I guess?

Comment (Yannig) [reply README.md]: Yes, you're right.

**Comment by Yannig** [comment]
@greenpau travis is now ok.

**Comment by greenpau** [comment]
@Yannig , please consider changing the commit message: `Big refactor of exporter`.

By and large, you are adding the ability to add metrics by way of a configuration file. The benefit of this is avoiding re-compiling the exporter every time you need to modify/fix a query or add a metric.

If it was my call, the commit message would be akin to "enhancement: add metrics via a configuration file"
 
On another note, when I am defining a metric like this, how do I know whether it is `gauge` or `counter` type? Where do I get `HELP` from? 

```
[[metric]]
context = "context_no_label"
request = "SELECT 1 as value_1, 2 as value_2 FROM DUAL"
metricsdesc = { value_1 = "Simple example returning always 1.", value_2 = "Same but returning always 2." }
```

**Comment by Yannig** [comment]
@greenpau commit message is now changed.

Actually, this exporter return only gauge. Maybe we can add a way to specify metric type.

Help field is generated using metricsdesc.

**Comment by greenpau** [comment]
> Actually, this exporter return only gauge. Maybe we can add a way to specify metric type.

@Yannig , I think it is reasonable, since some metrics are counters and some are gauges.

**Comment by Yannig** [comment]
@greenpau done

**Comment by greenpau** [comment]
> @greenpau done

@Yannig , in `.toml` files, where do you specify the `TYPE` for a metric?

**Comment by Yannig** [comment]
You have an example in **metric-dual-example.toml** and in **README.md**.

Here is an example:

```
# Can be counter or gauge (default)
metricstype = { value_1 = "counter" }
```

**Comment by greenpau** [comment]
@iamseth , this LGTM 👍

**Comment by Yannig** [comment]
Thanks :)

**Comment by Yannig** [comment]
@iamseth any comment?

**Comment by greenpau** [comment]
@urbanchef , I am a bystander here 😄 It is up to @Yannig and @iamseth

**Comment by Yannig** [comment]
@urbanchef Well actually it's up to @iamseth
I maintain my own fork and backport commits from this repository.
If you are interested, you can retrieve all this at the following location: https://github.com/Yannig/oracledb_exporter

**Comment by urbanchef** [comment]
@greenpau oops! :) Mixed you up.
@Yannig Thanks! Appreciate your effort!

**Comment by iamseth** [comment]
@Yannig sorry I don't really have time for this project or use Oracle anymore. I apologize for not responding sooner.

I've made you a collaborator on this repo if you want to handle the merges, etc. Alternatively I can change the README and redirect people to your fork.

**Comment by Yannig** [comment]
Thanks @iamseth

**Comment by Yannig** [comment]
@iamseth I have pushed my changes on **refactor** branch. I will test this to I will test this branch on my side. If I do not see anything particular to report and that you do not see any objection, I would like to make a version 0.2.

**Pull request closed without merging**